src/aapns/connection.py

```diff
@@ -0,0 +1,434 @@
+                    await self.bgw
+
+            if self.bgr:
+                self.bgr.cancel()
+                with suppress(CancelledError):
+                    await self.bgr
+
+            # at this point, we must release or cancel all pending requests
+            for sid, ch in self.channels.items():
+                if ch.fut and not ch.fut.done():
+                    ch.fut.set_exception(Closed(self.outcome))
+
+            self.w.close()
+            with suppress(ssl.SSLError):
+                await self.w.wait_closed()
+        finally:
+            self.closed = True
+
+    async def background_read(self):
+        try:
+            while not self.closed:
+                data = await self.r.read(2 ** 16)
+                if not data:
+                    break
+
+                for event in self.conn.receive_data(data):
+                    logger.debug("APN: %s", event)
+                    sid = getattr(event, "stream_id", 0)
+                    error = getattr(event, "error_code", None)
+                    ch = self.channels.get(sid)
+
+                    if isinstance(event, h2.events.RemoteSettingsChanged):
+                        m = event.changed_settings.get(
+                            h2.settings.SettingCodes.MAX_CONCURRENT_STREAMS
+                        )
+                        if m:
+                            self.max_concurrent_streams = m.new_value
+                    elif isinstance(event, h2.events.ConnectionTerminated):
+                        self.closing = True
+                        if not self.outcome:
+                            if event.additional_data:
+                                try:
+                                    self.outcome = json.loads(
+                                        event.additional_data.decode("utf-8")
+                                    )["reason"]
+                                except Exception:
+                                    self.outcome = str(self.additional_data[:100])
+                            else:
+                                self.outcome = str(event.error_code)
+                        logger.info("%s %s", self, self.outcome)
+                    elif not sid and error is not None:
+                        # FIXME break the connection,but give users a chance to complete
+                        logger.warning("Bad error %s", event)
+                        self.outcome = str(error)
+                        self.closing = True
+                    else:
+                        if isinstance(event, h2.events.DataReceived):
+                            # Stream flow control is responsibility of the channel.
+                            # Connection flow control is handled here.
+                            self.conn.increment_flow_control_window(
+                                event.flow_controlled_length
+                            )
+                        if ch:
+                            ch.ev.append(event)
```

Comment:
Am I correct in understanding that the reason for mixing h2 event handling here & in `post` is request timeout handling?

Reply:
Roughly it's like this:
* `background_read` handles connection events and dispatches streams to waiters
* `post` handles individual stream events
--
I think, I made it so for readability, to have neat, `async/await` **imperative** code.
It's true that timeouts are easier this way too 👍